### ISSUE TITLE
Sql: Allow array counting

### ISSUE THREAD

**Issue by eMerzh** [crate/crate#882]
Hi, 

it would be really great if we could do some basic arrays operations like array length, or array intersect (like the && operator in postgresql ie) ...

Thanks a lot for your work
##

**Comment by devoncrouse** [comment]
+1

**Comment by joemoe** [comment]
@eMerzh we added some array functions in the last time.
https://crate.io/docs/reference/en/latest/sql/scalar.html#array-functions
do you still need the count or overlap?

**Comment by eMerzh** [comment]
thanks @joemoe :) ... seems better, but how do you count the length of and array now?

**Comment by joemoe** [comment]
@eMerzh at the moment you can't.
the problem is that for the example you give the query will be either time consuming because the function needs to be applied on every row or we need to build a function based index.
as i interpreted your request it looked like you have growing arrays, each with the same elements. that's why array_intersects would also do its job. is this true?
this would be much faster.

**Comment by eMerzh** [comment]
yes like you have and array of email... you want to know every users with 2 emails... or smth like that...

**Comment by joemoe** [comment]
ok thanks. are you still using crate?

**Comment by eMerzh** [comment]
not currently ... ;-) so not top priority for now ;)

**Comment by diyism** [comment]
Backed this, https://www.bountysource.com/issues/2667489-sql-allow-array-counting

**Comment by diyism** [comment]
I need "select sum(array_count(array_matched(time_visit))) as total from statis_user_success where any(time_visit)>1481009355". Related to https://www.bountysource.com/issues/39611200-feature-request-support-array_matched-to-filter-elements-by-any-condition

**Comment by diyism** [comment]
After 3 years and 4 months, any progress?

**Comment by joemoe** [comment]
@diyism unfortunately not.

**Comment by diyism** [comment]
Maybe subquery can help:
select (select count(*) from unnest(tb1.array1)) from tb1

**Comment by diyism** [comment]
If we have group_concat():
`select char_length(regexp_replace(group_concat(tb1.array1)||',',  '[^,]*,', ',', 'g'))`
but we can only:
`select char_length(regexp_replace('ac,b,23'||',', '[^,]*,', ',', 'g'))`

**Comment by diyism** [comment]
Currently we can add another column:
```
insert into tb1 (key, dates, count_dates) values ('id_2', [1509000001], 1);
update tb1 set dates=array_cat(dates, [1509000401]),count_dates=count_dates+1 where key='id_2';
select count_dates from tb1 where key='id_2';
```
But if i insert duplicated "dates", the count_dates will be incorrect:
```
update tb1 set dates=array_cat(dates, [1509000401]),count_dates=count_dates+1 where key='id_2';
update tb1 set dates=array_cat(dates, [1509000401]),count_dates=count_dates+1 where key='id_2';
select count_dates from tb1 where key='id_2';
```

**Comment by diyism** [comment]
The CrateDB 2.3.0 has came, the array count() function is still not available.
I think you can only add an extra count index to all array fields to realize this count() function.

**Comment by mfussenegger** [comment]
We've added [array_upper](https://github.com/crate/crate/pull/7711) and
[array_length](https://github.com/crate/crate/pull/7743) which can be used for
the use-cases here.

They'll be available in 3.2 and can already be used in the latest nightly.

As already mentioned in this issue in some cases performance of using the
scalars in the where clause can be problematic. We've some [optimizations
pending](https://github.com/crate/crate/pull/7754) but not all cases can be
optimized efficiently.
